src/main/java/com/csc530/sat/DecisionDiagramNode.java

```diff
@@ -0,0 +1,190 @@
+        return new DecisionDiagramNode(primaryVariable, variables, trueBranch.not(),
+                falseBranch.not());
+    }
+
+    public boolean satisfies(final Map<SATVariable, Boolean> assignment) {
+        // This tree cannot possibly be satisfied if the assignment is empty
+        if (assignment.isEmpty()) {
+            return false;
+        }
+
+        // Or if it doesn't contain this node's variable!
+        if (!assignment.containsKey(primaryVariable)) {
+            return false;
+        }
+
+        DecisionDiagram branch = assignment.get(primaryVariable) ? trueBranch
+                : falseBranch;
+
+        // If the assignment does contain the variable, remove it from the
+        // assignment
+        // and recursively check the branch children
+        Map<SATVariable, Boolean> reducedAssignment = assignment.keySet().stream()
+                .filter(k -> !primaryVariable.equals(k))
+                .collect(Collectors.toMap(k -> k, assignment::get));
+
+        return branch.satisfies(reducedAssignment);
+    }
+
+    public DecisionDiagram assume(final SATVariable variable, final Boolean value) {
+        if (!variables.contains(variable)) {
+            return this;
+        }
+
+        if (primaryVariable.equals(variable)) {
+            return value ? trueBranch : falseBranch;
+        }
+
+        DecisionDiagram trueBranchAssumed = trueBranch.assume(variable, value);
+        DecisionDiagram falsedBranchAssumed = falseBranch.assume(variable, value);
+        if (canSimplify(trueBranchAssumed, falsedBranchAssumed)) {
+            return trueBranchAssumed;
+        }
+
+        return new DecisionDiagramNode(
+                primaryVariable, variables.stream().filter(v -> !v.equals(variable))
+                        .collect(Collectors.toSet()),
+                trueBranchAssumed, falsedBranchAssumed);
+    }
+
+    private boolean canSimplify(final DecisionDiagram d1, final DecisionDiagram d2) {
+        return d1 instanceof DecisionDiagramLeaf && d2 instanceof DecisionDiagramLeaf
+                && d1 == d2;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (o == this) {
+            return true;
+        }
+
+        if (o == null || !(o instanceof DecisionDiagramNode)) {
+            return false;
+        }
+
+        DecisionDiagramNode other = (DecisionDiagramNode) o;
+        return primaryVariable.equals(other.primaryVariable)
+                && variables.equals(other.variables)
+                && trueBranch.equals(other.trueBranch)
+                && falseBranch.equals(other.falseBranch);
+    }
+
+    /**
+     * Provides a JSON string representation to easily use in debugging.
+     */
+    @Override
+    public String toString() {
+        return String.format("{\"%s\": {\"true\": %s, \"false\": %s}}", primaryVariable,
+                trueBranch, falseBranch);
+    }
+
+    @Override
+    public Set<Map<SATVariable, Boolean>> satisifyAll() {
```

Comment:
I can make a method latter that just satisfies one  this is just for trees that have a few number of solutions